POLIO-1655 adapt stock backend

quang-le opened the pull request:
Backend was ignoring usable vials on some incident reports

Related JIRA tickets : POLIO-1655

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
NA

## Changes

- LOSSES, STEALING, BROKEN and RETURN incident types now check only usable vials and don't update unusable vials
- Commented out OTHER, since it's still being specced
-  Fixed an (unrelated) failing test for vaccine authorizations
- Update some logic:
    - vials received: only take into account vials from arrival reports. Exclude those found in physical inventory
    - vials used: aggregate vials from form A outgoing movement i.o using the difference between received and usable vials
    - Add methods to VaccineStockCalculator to make these calculations
    - Refactor existing methods (mostly because the new methods just extract some code from the old ones)
- Update and comment tests

## How to test

Go to vaccine stock management
Add one of the listed incident report types to an existing stock
Save
Go back to stock list
The incident report should appear, in usable vials table, with the amount in vials OUT. It should not appear in the unusable vials table

## Print screen / video

https://github.com/user-attachments/assets/c4cb3049-57a1-4fe1-a69a-328f4613e3a7


## Notes

merges into https://github.com/BLSQ/iaso/pull/1575

quang-le committed:
POLIO-1655: replace check on unusable vials where needed

- LOSSES, STEALING, BROKEN and RETURN incident types now check only usable vials and don't update unusable vials
- Commented out OTHER, since it's still being specced

quang-le committed:
POLIO-1655: fix failing test on vaccine authorizations

kemar commented on plugins/polio/api/vaccines/stock_management.py:
```diff
@@ -263,7 +275,7 @@ def get_list_of_unusable_vials(self):
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.VVM_REACHED_DISCARD_POINT
                 or report.stock_correction == IncidentReport.StockCorrectionChoices.UNREADABLE_LABEL
-                or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
+                # or report.stock_correction == IncidentReport.StockCorrectionChoices.OTHER
```
~~`OTHER` looks commented everywhere. Is this OK?~~

Juste read the PR's desc:

> Commented out OTHER, since it's still being specced

So this looks OK.

kemar replied:
Maybe add a comment about that so that a new reader will understand why this is commented?

quang-le replied:
It think this will be updated in the base branch

quang-le committed:
POLIO-1655: fix vaccine stock calculations
- vials received: only take into account vials from arrival reports. Exclude those found in physicla inventory
- vials used: aggregate vials from form A outgoing movement i.o using the difference between received and usable vials
- Add methods to VaccineStockCalculator to make these calculations
- Refactor existing methods (mostly because the new methods just extract some code from the old ones)
- Update and comment tests

kemar commented on plugins/polio/api/vaccines/stock_management.py:
```diff
@@ -117,7 +119,10 @@ def get_total_of_unusable_vials(self):
-    def get_list_of_usable_vials(self):
+    def get_list_of_vaccines_received(self):
+        """
+        Vaccines received are only those linked to an arrival report. We exclude those found e;g during physical inventory
```
```suggestion
        Vaccines received are only those linked to an arrival report. We exclude those found e.g. during physical inventory
```

quang-le committed:
Merge branch 'POLIO-1636-incident-reports' into POLIO-1655_adapt_stock-backend

kemar commented on plugins/polio/api/vaccines/stock_management.py:
```diff
@@ -242,6 +257,17 @@ def get_list_of_unusable_vials(self):
+        return results
+
+    def get_list_of_unusable_vials(self):
+        # First get the used viald
```
```suggestion
        # First get the used vials
```

kemar commented on plugins/polio/tests/test_vaccine_authorizations.py:
```diff
@@ -658,10 +659,8 @@ def test_expired_vaccine_authorizations_email_alert(self):
         response = expired_vaccine_authorizations_email_alert(vaccine_auths, mailing_list)
-
-        page_url = f"example.com/dashboard/polio/vaccinemodule/nopv2authorisation/accountId/{self.team.project.account.id}/order/-current_expiration_date/pageSize/20/page/1"
+        page_url = f"{Site.objects.get_current()}/dashboard/polio/vaccinemodule/nopv2authorisation/accountId/{self.team.project.account.id}/order/-current_expiration_date/pageSize/20/page/1"
```
I'd rather use `settings.DNS_DOMAIN` since we decided to not rely on `current_site.domain` in #1378.

kemar reviewed (approved):
LGTM
Some minor non-blocking feedback.

quang-le committed:
POLIO-1655: black formatting

quang-le committed:
Update plugins/polio/api/vaccines/stock_management.py

Co-authored-by: Marc Hertzog <marc@marcarea.com>

quang-le committed:
Update plugins/polio/api/vaccines/stock_management.py

Co-authored-by: Marc Hertzog <marc@marcarea.com>

quang-le committed:
POLIO-1655: use settings.DNS_DOMAIN for vaccine authorizations emails

quang-le committed:
Merge branch 'POLIO-1655_adapt_stock-backend' of https://github.com/BLSQ/iaso into POLIO-1655_adapt_stock-backend
